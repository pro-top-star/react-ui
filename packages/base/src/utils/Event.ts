import { ReactNode } from 'react';
import { HTMLEvent } from '../interfaces/HTMLEvent';

export interface Parameters {
  [key: string]: number | string | object | boolean | Parameters;
}

export class Event {
  readonly source: ReactNode;
  readonly originalEvent: HTMLEvent;
  readonly parameters: any;
  readonly htmlSource: EventTarget;

  /**
   * Event Constructor
   * @param source {ReactNode} The React Element instance
   * @param originalEvent {HTMLEvent} original HTML Event that triggered this event
   * @param parameters {Parameters} event payload
   */
  constructor(source: ReactNode, originalEvent: HTMLEvent, parameters: Parameters = {}) {
    this.source = source;
    this.originalEvent = originalEvent;
    this.parameters = parameters;
    this.htmlSource = originalEvent.target;
  }

  /**
   * Static initializer for an Event
   * @param source {ReactNode} The React Element instance
   * @param originalEvent {HTMLEvent} original HTML Event that triggered this event
   * @param parameters {Parameters} event payload
   */
  public static of(source: ReactNode, originalEvent: HTMLEvent, parameters: Parameters = {}) {
    return new Event(source, originalEvent, parameters);
  }

  public getSource(): ReactNode {
    return this.source;
  }

  public getOriginalEvent(): HTMLEvent {
    return this.originalEvent;
  }

  public getParameters(): Parameters {
    return this.parameters;
  }

  public getParameter(parameter: string): any {
    return this.parameters[parameter];
  }

  public getHtmlSourceElement(): EventTarget {
    return this.htmlSource;
  }

  public cancel(): void {
    this.originalEvent.stopPropagation();
  }

  public preventDefault(): void {
    this.originalEvent.preventDefault();
  }
}
