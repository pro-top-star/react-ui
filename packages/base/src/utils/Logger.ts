/* eslint-disable no-console,no-unused-expressions */

const pad0 = (i: any, w: number) => `000${String(i)}`.slice(-w);

export enum LOG_LEVEL {
  FATAL = 0,
  ERROR = 1,
  WARNING = 2,
  INFO = 3,
  DEBUG = 4,
  TRACE = 5
}

class DeviceLogger {
  public defaultComponent: string;
  public sWindowName = window.top === window ? '' : `[${window.location.pathname.split('/').slice(-1)[0]}] `;

  constructor(defaultComponent = 'DEVICE') {
    this.log(LOG_LEVEL.INFO, 'Device API logging initialized');
    this.defaultComponent = defaultComponent;
    return this;
  }

  public for(defaultComponent: string) {
    return new DeviceLogger(defaultComponent);
  }

  public log(iLevel: LOG_LEVEL, sMessage: string, sComponent?: string) {
    sComponent = sComponent || this.defaultComponent || '';
    const oNow = new Date();
    const oLogEntry = {
      time: `${pad0(oNow.getHours(), 2)}:${pad0(oNow.getMinutes(), 2)}:${pad0(oNow.getSeconds(), 2)}`,
      date: `${pad0(oNow.getFullYear(), 4)}-${pad0(oNow.getMonth() + 1, 2)}-${pad0(oNow.getDate(), 2)}`,
      timestamp: oNow.getTime(),
      level: iLevel,
      message: sMessage || '',
      component: sComponent || ''
    };
    if (window.console) {
      const logText = `${oLogEntry.date} ${oLogEntry.time} ${this.sWindowName}${oLogEntry.message} - ${oLogEntry.component}`;
      switch (iLevel) {
        case LOG_LEVEL.FATAL:
        case LOG_LEVEL.ERROR:
          console.error(logText);
          break;
        case LOG_LEVEL.WARNING:
          console.warn(logText);
          break;
        case LOG_LEVEL.INFO:
          console.info ? console.info(logText) : console.log(logText);
          break; // info not available in iOS simulator
        case LOG_LEVEL.DEBUG:
          console.debug ? console.debug(logText) : console.log(logText);
          break; // debug not available in IE, fallback to log
        case LOG_LEVEL.TRACE:
          console.trace ? console.trace(logText) : console.log(logText);
          break; // trace not available in IE, fallback to log
      }
    }
    return oLogEntry;
  }

  //  Public API
  public fatal(msg: string, component?: string) {
    return this.log(LOG_LEVEL.FATAL, msg, component);
  }

  public error(msg: string, component?: string) {
    return this.log(LOG_LEVEL.ERROR, msg, component);
  }

  public warning(msg: string, component?: string) {
    return this.log(LOG_LEVEL.WARNING, msg, component);
  }

  public info(msg: string, component?: string) {
    return this.log(LOG_LEVEL.INFO, msg, component);
  }

  public debug(msg: string, component?: string) {
    return this.log(LOG_LEVEL.DEBUG, msg, component);
  }

  public trace(msg: string, component?: string) {
    return this.log(LOG_LEVEL.TRACE, msg, component);
  }
}

export const Logger = new DeviceLogger();
