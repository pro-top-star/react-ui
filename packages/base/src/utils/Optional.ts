/* eslint-disable id-blacklist */
/**
 * Optionals for Javascript Objects
 * Inspired by Java Optionals
 */
class Optional {
  private readonly object: object;
  private path: string;

  /**
   * Creates a new Optional
   * @param {Object} object The Object to validate
   * @param {String} path Path to search for in object. Separated by dots.
   *                      Example: 'foo.bar'
   * @constructor
   */
  constructor(object: object, path: string) {
    this.object = object;
    this.path = path;
  }

  /**
   * Creates a new Optional
   * @param {Object} object The Object to validate
   * @param {String} path Path to search for in object. Separated by dots.
   *                      Example: 'foo.bar'
   * @constructor
   */
  static of(object: object, path: string) {
    return new Optional(object, path);
  }

  /**
   * Get The Value of the Optional.
   * Do not use the this method directly!
   * @return {any} Resolved Value
   * @private
   */
  get() {
    return this.path.split('.').reduce((acc, val) => {
      if (acc === undefined || acc === null) {
        return acc;
      }
      if (acc.hasOwnProperty(val)) {
        const newVal = acc[val];
        if (newVal !== undefined && newVal !== null) {
          return newVal;
        }
        return newVal;
      }
      return undefined;
    }, this.object);
  }

  /**
   * Return true if there is a value present, otherwise false.
   * @returns {Boolean} true if there is a value present, otherwise false
   */
  isPresent() {
    let obj: any = this.object;
    return this.path.split('.').every((x) => {
      if (typeof obj !== 'object' || obj === null || obj === undefined || !(x in obj)) return false;
      obj = obj[x];
      return true;
    });
  }

  /**
   * If a value is present, invoke the specified consumer with the value, otherwise do nothing.
   * @param {Function} fn block to be executed if a value is present
   * @returns {any} value The executed fn if present, else null
   */
  ifPresent(fn: (value: any) => void) {
    if (this.isPresent()) {
      return fn(this.get());
    }
    return null;
  }

  /**
   * Return the value if present, otherwise return other.
   * @param val {any} the value to be returned if there is no value present, may be null
   * @returns {any} value the value, if present, otherwise other
   */
  orElse(val: any) {
    return this.isPresent() ? this.get() : val;
  }

  /**
   * Return the contained value, if present, otherwise throw an exception to be created by the provided supplier.
   * @param {ExceptionSupplier} exceptionSupplier The supplier which will return the exception to be thrown
   * @returns {any} the present value
   * @throws {Exception} ex - if there is no value present
   *
   */
  orElseThrow(exceptionSupplier: () => Error) {
    if (!this.isPresent()) {
      throw exceptionSupplier();
    }
    return this.get();
  }
}

export { Optional };
