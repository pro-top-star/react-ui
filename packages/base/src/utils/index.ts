import { UIEvent } from 'react';

export const deprecationNotice = (component: string, message: string) => {
  if (process.env.NODE_ENV === 'development') {
    const value = `*** ui5-webcomponents-react Deprecation Notice - ${component} ***\n`;
    // eslint-disable-next-line no-console
    console.warn(`${value}${message}`);
  }
};

export const getScrollBarWidth = () => {
  const inner = document.createElement('p');
  inner.style.width = '100%';
  inner.style.height = '200px';

  const outer = document.createElement('div');
  outer.style.position = 'absolute';
  outer.style.top = '0px';
  outer.style.left = '0px';
  outer.style.visibility = 'hidden';
  outer.style.width = '200px';
  outer.style.height = '150px';
  outer.style.overflow = 'hidden';
  outer.appendChild(inner);

  document.body.appendChild(outer);
  const w1 = inner.offsetWidth;
  outer.style.overflow = 'scroll';
  let w2 = inner.offsetWidth;

  if (w1 === w2) {
    w2 = outer.clientWidth;
  }

  document.body.removeChild(outer);
  return w1 - w2;
};

export const enrichEventWithDetails = <T = {}>(event: UIEvent, payload: T = {} as any) => {
  if (event.hasOwnProperty('persist')) {
    // if there is a persist method, it's an SyntheticEvent so we need to persist it
    event.persist();
  }

  const shouldCreateNewDetails =
    event.detail === null || event.detail === undefined || typeof event.detail !== 'object';
  Object.defineProperty(event, 'detail', {
    value: shouldCreateNewDetails ? {} : event.detail,
    writable: true,
    configurable: true
  });
  Object.assign(event.detail, payload);
  return (event as unknown) as CustomEvent<T>;
};
