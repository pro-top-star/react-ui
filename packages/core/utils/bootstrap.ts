import { ExtendedWindow } from '../interfaces/ExtendedWindow';
import 'core-js/modules/es6.object.assign';
import 'core-js/modules/es7.object.values';

if (!!(window as ExtendedWindow).Fiori4React === false) Object.assign(window, { Fiori4React: {} });
if (!!(window as ExtendedWindow).Fiori4React.__SECRET_INTERNALS_DO_NOT_USE === false) {
  Object.assign((window as ExtendedWindow).Fiori4React, {
    __SECRET_INTERNALS_DO_NOT_USE: {}
  });
}
