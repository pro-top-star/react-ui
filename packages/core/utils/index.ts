export * from './Device';
export * from './isNumeric';
