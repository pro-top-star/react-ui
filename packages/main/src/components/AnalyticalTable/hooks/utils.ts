export const makeTemplateColumns = (columns, resizedColumns) => {
  return columns
    .map((column) => {
      if (resizedColumns[column.id]) {
        return `${resizedColumns[column.id]}px`;
      }
      return `minmax(${column.minWidth}px, ${column.width || '1fr'})`;
    })
    .join(' ');
};
