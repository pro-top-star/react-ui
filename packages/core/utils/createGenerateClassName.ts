import './bootstrap';
import { ExtendedWindow } from '../interfaces/ExtendedWindow';

const env = process.env.NODE_ENV;
const prodClassPrefix = 'f4r';

const extractClassNamePrefix = (className) => {
  const match = className.match && className.match(/WithStyles\((.*)\)/);
  return (match && match[1].replace(/\$/g, '')) || className;
};

export const createGenerateClassName = () => {
  const defaultPrefix = env === 'production' ? prodClassPrefix : '';
  if (!!(window as ExtendedWindow).Fiori4React.__SECRET_INTERNALS_DO_NOT_USE.jssRuleCounter === false) {
    Object.assign((window as ExtendedWindow).Fiori4React.__SECRET_INTERNALS_DO_NOT_USE, { jssRuleCounter: 0 });
  }
  const internals = (window as ExtendedWindow).Fiori4React.__SECRET_INTERNALS_DO_NOT_USE;
  return (rule, sheet) => {
    internals.jssRuleCounter++;

    let prefix = defaultPrefix;
    let jssId = '';

    if (sheet) {
      prefix = env === 'production' ? defaultPrefix : extractClassNamePrefix(sheet.options.classNamePrefix);
      if (sheet.options.jss.id != null) jssId += sheet.options.jss.id;
    }

    if (env === 'production') {
      return `${prefix}${jssId}${internals.jssRuleCounter}`;
    }

    return `${prefix}-${rule.key}-${jssId && `-${jssId}`}-${internals.jssRuleCounter}`;
  };
};
