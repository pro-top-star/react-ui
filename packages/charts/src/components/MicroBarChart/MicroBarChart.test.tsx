import { renderThemedComponent } from '@shared/tests/utils';
import * as React from 'react';
import { MicroBarChart } from './index';

const dataset = [
  { value: 10, label: 'Bar Number One' },
  { value: 100, label: 'Bar Number Two' }
];

describe('Micro Bar Chart', () => {
  test('Renders with data', () => {
    expect(renderThemedComponent(<MicroBarChart dataset={dataset} />)).toMatchSnapshot();
  });

  test('custom colors', () => {
    expect(renderThemedComponent(<MicroBarChart dataset={dataset} colors={['#f0ab00']} />)).toMatchSnapshot();
  });

  test('element selection', () => {
    expect(renderThemedComponent(<MicroBarChart dataset={dataset} visibleDatasetCount={1} />)).toMatchSnapshot();
  });

  test('maxWidth selection', () => {
    expect(renderThemedComponent(<MicroBarChart dataset={dataset} maxWidth={'200px'} />)).toMatchSnapshot();
  });
});
