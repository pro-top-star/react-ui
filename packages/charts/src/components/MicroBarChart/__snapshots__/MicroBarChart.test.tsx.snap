// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Micro Bar Chart Renders with data 1`] = `
<div
  class="MicroBarChart--container- undefined"
>
  <div>
    <div
      class="MicroBarChart--labelContainer-"
    >
      <span
        class="MicroBarChart--label-"
        title="Bar Number One"
      >
        Bar Number One
      </span>
      <span
        class="MicroBarChart--text-"
        style="font-size:12px"
        title="10"
      >
        10
      </span>
    </div>
    <div
      class="MicroBarChart--valueContainer-"
    >
      <div
        class="MicroBarChart--valueBar-"
        style="width:10%;background-color:#0f828f"
      />
      <div
        class="MicroBarChart--fillUp-"
      />
    </div>
  </div>
  <div>
    <div
      class="MicroBarChart--labelContainer-"
    >
      <span
        class="MicroBarChart--label-"
        title="Bar Number Two"
      >
        Bar Number Two
      </span>
      <span
        class="MicroBarChart--text-"
        style="font-size:12px"
        title="100"
      >
        100
      </span>
    </div>
    <div
      class="MicroBarChart--valueContainer-"
    >
      <div
        class="MicroBarChart--valueBar-"
        style="width:100%;background-color:#5ac2ce"
      />
      <div
        class="MicroBarChart--fillUp-"
      />
    </div>
  </div>
</div>
`;

exports[`Micro Bar Chart custom colors 1`] = `
<div
  class="MicroBarChart--container- undefined"
>
  <div>
    <div
      class="MicroBarChart--labelContainer-"
    >
      <span
        class="MicroBarChart--label-"
        title="Bar Number One"
      >
        Bar Number One
      </span>
      <span
        class="MicroBarChart--text-"
        style="font-size:12px"
        title="10"
      >
        10
      </span>
    </div>
    <div
      class="MicroBarChart--valueContainer-"
    >
      <div
        class="MicroBarChart--valueBar-"
        style="width:10%;background-color:#f0ab00"
      />
      <div
        class="MicroBarChart--fillUp-"
      />
    </div>
  </div>
  <div>
    <div
      class="MicroBarChart--labelContainer-"
    >
      <span
        class="MicroBarChart--label-"
        title="Bar Number Two"
      >
        Bar Number Two
      </span>
      <span
        class="MicroBarChart--text-"
        style="font-size:12px"
        title="100"
      >
        100
      </span>
    </div>
    <div
      class="MicroBarChart--valueContainer-"
    >
      <div
        class="MicroBarChart--valueBar-"
        style="width:100%"
      />
      <div
        class="MicroBarChart--fillUp-"
      />
    </div>
  </div>
</div>
`;

exports[`Micro Bar Chart element selection 1`] = `
<div
  class="MicroBarChart--container- undefined"
>
  <div>
    <div
      class="MicroBarChart--labelContainer-"
    >
      <span
        class="MicroBarChart--label-"
        title="Bar Number One"
      >
        Bar Number One
      </span>
      <span
        class="MicroBarChart--text-"
        style="font-size:12px"
        title="10"
      >
        10
      </span>
    </div>
    <div
      class="MicroBarChart--valueContainer-"
    >
      <div
        class="MicroBarChart--valueBar-"
        style="width:10%;background-color:#0f828f"
      />
      <div
        class="MicroBarChart--fillUp-"
      />
    </div>
  </div>
</div>
`;

exports[`Micro Bar Chart maxWidth selection 1`] = `
<div
  class="MicroBarChart--container- undefined"
  style="max-width:200px"
>
  <div>
    <div
      class="MicroBarChart--labelContainer-"
    >
      <span
        class="MicroBarChart--label-"
        title="Bar Number One"
      >
        Bar Number One
      </span>
      <span
        class="MicroBarChart--text-"
        style="font-size:12px"
        title="10"
      >
        10
      </span>
    </div>
    <div
      class="MicroBarChart--valueContainer-"
    >
      <div
        class="MicroBarChart--valueBar-"
        style="width:10%;background-color:#0f828f"
      />
      <div
        class="MicroBarChart--fillUp-"
      />
    </div>
  </div>
  <div>
    <div
      class="MicroBarChart--labelContainer-"
    >
      <span
        class="MicroBarChart--label-"
        title="Bar Number Two"
      >
        Bar Number Two
      </span>
      <span
        class="MicroBarChart--text-"
        style="font-size:12px"
        title="100"
      >
        100
      </span>
    </div>
    <div
      class="MicroBarChart--valueContainer-"
    >
      <div
        class="MicroBarChart--valueBar-"
        style="width:100%;background-color:#5ac2ce"
      />
      <div
        class="MicroBarChart--fillUp-"
      />
    </div>
  </div>
</div>
`;
